template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -291,6 +311,52 @@ public String getStackVersion() {
         return cmTemplate.getCdhVersion();
     }
 
+    @Override
+    public Map<String, Map<String, ServiceAttributes>> getHostGroupBasedServiceAttributes() {
+        // The only attributes required at the moment relate to the YARN NodeManager.
+        // This functionality can be moved to an interface / separated if additional components
+        // need this functionality.
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponentsWithYarnNMs = new HashMap<>();
+        for (Entry<String, Set<ServiceComponent>> entry : hgToNonGwServiceComponents.entrySet()) {
+            for (ServiceComponent sc : entry.getValue()) {
+                if (sc.getService().equalsIgnoreCase(YarnRoles.YARN) && sc.getComponent()
+                        .equalsIgnoreCase(YarnRoles.NODEMANAGER)) {
+                    hgToNonGwServiceComponentsWithYarnNMs.put(entry.getKey(), entry.getValue());
+                    break;
+                }
+            }
+        }
+        // We have the hostGroups without GATEWAY components with YARN NodeManagers
```

Comment:
instead of comment could you use variable and methods name reflecting the same?

Reply:
Think this is very specific to CB code style? I think comments in the code can be quite helpful.

Reply:
comments tends to be outdated, misleading. Usually indicates the code is not "clean", doesn't speak for itself which could be improved by refactoring, like breaking into methods with good names, etc
We generally avoid them as much as possible and leave comments where it's really necessary.